engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java

```diff
@@ -0,0 +1,178 @@
+
+    // when
+    final CompletableFuture<Void> callbackFuture = new CompletableFuture<>();
+    final Runnable callback = () -> callbackFuture.complete(null);
+    monitor.addOnIdleCallback(callback);
+
+    // then
+    callbackFuture.get(1L, TimeUnit.SECONDS);
+    assertThat(callbackFuture).isCompleted();
+  }
+
+  @Test
+  void testCallbackIsCalledWhenEngineIsAlreadyBusy()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    // given
+    final InMemoryLogStorage logStorage = mock(InMemoryLogStorage.class);
+    final TestLogStreamReader logStreamReader = new TestLogStreamReader();
+    final EngineStateMonitor monitor = new EngineStateMonitor(logStorage, logStreamReader);
+    changeToBusyState(monitor, logStreamReader);
+    // We need to set the position of the reader high enough the so hasNext() resolves to false. If
+    // we don't do this adding the callback will make the engine idle because of the
+    // forwardToLastEvent() method in the EngineStateMonitor, and the callback will never get called
+    logStreamReader.setPosition(100L);
+
+    // when
+    final CompletableFuture<Void> callbackFuture = new CompletableFuture<>();
+    final Runnable callback = () -> callbackFuture.complete(null);
+    monitor.addOnProcessingCallback(callback);
+
+    // then
+    callbackFuture.get(1L, TimeUnit.SECONDS);
+    assertThat(callbackFuture).isCompleted();
+  }
+
+  @Test
+  void testCallbackIsCalledWhenBusyStateIsReached()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    // given
+    final InMemoryLogStorage logStorage = mock(InMemoryLogStorage.class);
+    final TestLogStreamReader logStreamReader = new TestLogStreamReader();
+    final EngineStateMonitor monitor = new EngineStateMonitor(logStorage, logStreamReader);
+
+    final CompletableFuture<Void> callbackFuture = new CompletableFuture<>();
+    final Runnable callback = () -> callbackFuture.complete(null);
+    monitor.addOnProcessingCallback(callback);
+
+    // when
+    changeToBusyState(monitor, logStreamReader);
+
+    // then
+    callbackFuture.get(1L, TimeUnit.SECONDS);
+    assertThat(callbackFuture).isCompleted();
+  }
+
+  private void changeToIdleState(final EngineStateMonitor monitor, final TestLogStreamReader reader) {
+    // We use onCommit here because it is an easy way to trigger the EngineStateMonitor to check the
+    // engine state and trigger the callbacks
+    reader.setPosition(reader.getLastEventPosition());
+    monitor.onCommit();
+  }
+
+  private void changeToBusyState(final EngineStateMonitor monitor, final TestLogStreamReader reader) {
+    // We use onReplayed here because it is an easy way to update the lastProcessedEventPosition of
+    // the EngineStateMonitor
+    final long position = reader.getLastEventPosition() + 1;
+    monitor.onReplayed(position, -1L);
+    reader.setLastEventPosition(position);
+  }
+
+  private class TestLogStreamReader implements LogStreamReader {
```

Comment:
:wrench: I think you could just use mocks for this. But I don't think this too verbose, either.

Reply:
I'd rather stick with this solution in that case. I think mocking is gonna make the test more confusing.